wicket-dbg/src/server.rs

```diff
@@ -0,0 +1,95 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::{Cmd, Runner};
+use std::fs::File;
+use std::net::{TcpListener, TcpStream, ToSocketAddrs};
+use std::path::PathBuf;
+
+/// The server used to handle wicket-dbg commands, run them, and return the
+/// response to the client.
+///
+/// We only allow one client to connect at a time because:
+///   1. The client is a human trying to debug
+///   2. There is only one screen to display the wicket UI on
+pub struct Server {
+    recording_path: Option<PathBuf>,
+    runner: Runner,
+}
+
+impl Server {
+    pub fn run<A: ToSocketAddrs>(
+        addr: A,
+        runner: Runner,
+    ) -> anyhow::Result<()> {
+        let listener = TcpListener::bind(addr)?;
+
+        let mut server = Server { recording_path: None, runner };
+
+        // accept connections and process them serially
+        for stream in listener.incoming() {
+            server.handle_client(stream?)?;
+        }
+        Ok(())
+    }
+
+    fn handle_client(&mut self, mut stream: TcpStream) -> anyhow::Result<()> {
+        loop {
+            let cmd: Cmd = ciborium::de::from_reader(&mut stream)?;
+            self.handle(cmd, &mut stream)?;
+        }
+    }
+
+    fn handle(
+        &mut self,
+        cmd: Cmd,
+        stream: &mut TcpStream,
```

Comment:
I think the stream here and elsewhere needs some sort of buffering around it.

I'd also express a preference for using async code here, though I'll let you decide whether you want to do that or not. (It's ok to use cbor to convert the data into an in-memory chunk and then write that out over the wire.)

Reply:
> I think the stream here and elsewhere needs some sort of buffering around it.

I don't think that's true. We only send one command at a time, and the operations are synchronous. We're not trying to reduce syscalls here, we want to send with as low-latency as possible. This is also why I turned off nagle on the socket, so that the cli responds quickly.

> 
> I'd also express a preference for using async code here, though I'll let you decide whether you want to do that or not. (It's ok to use cbor to convert the data into an in-memory chunk and then write that out over the wire.)

Reply:
On second thought, I'm switching over to async for the server in a follow up as I add more functionality. Specifically, the ability to pause playback and get back the current event index paused with a oneshot channel is a leading reason.

Reply:
> I don't think that's true.

So I very recently fixed an issue with dropshot's test harness which was incredibly slow because `serde_json` was writing one character at a time: https://github.com/oxidecomputer/dropshot/pull/539

I'm not entirely sure what ciborium does, but adding a little bit of buffering before flushing it out immediately would make sense I think without causing any sort of significant delay.

Reply:
> > I don't think that's true.
> 
> So I very recently fixed an issue with dropshot's test harness which was incredibly slow because `serde_json` was writing one character at a time: [oxidecomputer/dropshot#539](https://github.com/oxidecomputer/dropshot/pull/539)
> 

Wow, that's some crazy pathological behavior. You literally can't get worse 😆  😭. And yeah, I don't think a buffer really hurts even if it adds a touch of latency, due to the fact we are dealing with human timescales here. I already have code that buffers and uses tokio, so you I guess I decided you were right!

> I'm not entirely sure what ciborium does, but adding a little bit of buffering before flushing it out immediately would make sense I think without causing any sort of significant delay.

I assumed it would behave sanely and dump as much as it could in a single write, but that is clearly not always a good assumption. Thanks for the pointer. This has been a learning experience lol.